utils/zerovec/src/map/mod.rs

```diff
@@ -0,0 +1,273 @@
+    ///
+    /// ```rust
+    /// use zerovec::ZeroMap;
+    ///
+    /// let mut map = ZeroMap::new();
+    /// map.insert(1, "one".to_owned());
+    /// map.insert(2, "two".to_owned());
+    /// assert_eq!(map.remove(&1), Some("one".to_owned()));
+    /// assert_eq!(map.get(&1), None);
+    /// ```
+    pub fn remove(&mut self, key: &K::NeedleType) -> Option<V> {
+        let idx = self.keys.binary_search(key).ok()?;
+        self.keys.remove(idx);
+        Some(self.values.remove(idx))
+    }
+
+    /// Appends `value` with `key` to the end of the underlying vector, returning
+    /// `key` and `value` _if it failed_. Useful for extending with an existing
+    /// sorted list.
+    /// ```rust
+    /// use zerovec::ZeroMap;
+    ///
+    /// let mut map = ZeroMap::new();
+    /// assert!(map.try_append(1, "uno".to_owned()).is_none());
+    /// assert!(map.try_append(3, "tres".to_owned()).is_none());
+    ///
+    /// let unsuccessful = map.try_append(3, "tres-updated".to_owned());
+    /// assert!(unsuccessful.is_some(), "append duplicate of last key");
+    ///
+    /// let unsuccessful = map.try_append(2, "dos".to_owned());
+    /// assert!(unsuccessful.is_some(), "append out of order");
+    ///
+    /// assert_eq!(map.get(&1), Some("uno"));
+    ///
+    /// // contains the original value for the key: 3
+    /// assert_eq!(map.get(&3), Some("tres"));
+    ///
+    /// // not appended since it wasn't in order
+    /// assert_eq!(map.get(&2), None);
+    /// ```
+    #[must_use]
+    pub fn try_append(&mut self, key: K, value: V) -> Option<(K, V)> {
+        if self.keys.len() != 0 {
+            if let Some(last) = self.keys.get(self.keys.len() - 1) {
+                if key.cmp_get(last) != Ordering::Greater {
+                    return Some((key, value));
+                }
+            }
+        }
+
+        self.keys.push(key);
+        self.values.push(value);
+        None
+    }
+
+    /// Produce an ordered iterator over key-value pairs
+    pub fn iter<'b>(
+        &'b self,
+    ) -> impl Iterator<
+        Item = (
+            &'b <K as ZeroMapKV<'a>>::GetType,
+            &'b <V as ZeroMapKV<'a>>::GetType,
+        ),
+    > {
+        (0..self.keys.len())
+            .map(move |idx| (self.keys.get(idx).unwrap(), self.values.get(idx).unwrap()))
+    }
+
+    /// Produce an ordered iterator over keys
+    pub fn iter_keys<'b>(&'b self) -> impl Iterator<Item = &'b <K as ZeroMapKV<'a>>::GetType> {
+        (0..self.keys.len()).map(move |idx| self.keys.get(idx).unwrap())
+    }
+
+    /// Produce an ordered iterator over keys
+    pub fn iter_values<'b>(&'b self) -> impl Iterator<Item = &'b <V as ZeroMapKV<'a>>::GetType> {
+        (0..self.values.len()).map(move |idx| self.values.get(idx).unwrap())
+    }
+}
+
+impl<'a, K, V> ZeroMap<'a, K, V>
+where
+    K: ZeroMapKV<'a>,
+    V: ZeroMapKV<'a, Container = ZeroVec<'a, V>>,
+    V: AsULE + Copy,
+    V::ULE: 'static,
```

Comment:
Question, here and below: Why do we need the `'static` constraint?

Reply:
Should we add the `'static` to the ULE trait, then?  Like `trait ULE: Sized + 'static`

Reply:
Sure